src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -1,138 +1,433 @@
+                          name={item.name}
+                          key={item.id}
+                          {...(onMixedStateChange !== undefined
+                            ? {
+                                isChecked: isAllChecked(id, item),
+                                isIndeterminate: isIndeterminate(id, item),
+                                onChange: onMixedStateChange,
+                                isDisabled: updateDisabledState(id, item),
+                              }
+                            : {
+                                isChecked: isChecked(id, item.id),
+                                isDisabled: updateDisabledState(id, item),
+                                onChange: onChange,
+                              })}
+                        />
+                        {item.children.map((childItem) => (
+                          <Checkbox
+                            key={childItem.id}
+                            marginInlineStart="0"
+                            __css={styles.menuChildren}
+                            id={childItem.id}
+                            labelText={childItem.name}
+                            name={childItem.name}
+                            isDisabled={childItem.isDisabled}
+                            isChecked={isChecked(id, childItem.id)}
+                            onChange={onChange}
+                          />
+                        ))}
+                      </>
+                    ) : (
+                      <Checkbox
+                        id={item.id}
+                        labelText={item.name}
+                        name={item.name}
+                        isDisabled={item.isDisabled}
+                        isChecked={isChecked(id, item.id)}
+                        onChange={onChange}
+                        key={item.id}
+                      />
+                    )}
+                  </React.Fragment>
+                ))}
+              </CheckboxGroup>
+              {showViewLabel()}
+            </Box>
+          </>
         );
-      }
+      };
 
-      if (type === "dialog") {
-        const dialogOnChange = onChange as DialogOnChange;
+      const showViewLabel = () => {
+        if (listOverflow === "expand" && items.length >= defaultItemsVisible) {
+          return (
+            <Box
+              onClick={() => viewAllItems()}
+              __css={styles.viewAllButton}
+              id="view-all-text-btn"
+              marginTop="s"
+            >
+              {viewAllLabel}
+            </Box>
+          );
+        }
+      };
 
+      const itemsNotFound = () => {
+        if (itemsList.length === 0) {
+          return (
+            <Box id="items-not-found-text-id" marginTop="xs">
+              No options found
+            </Box>
+          );
+        }
+      };
+
+      const viewAllItems = () => {
+        setViewAllLabel(viewAllLabel === "View all" ? "View less" : "View all");
```

Comment:
These `set....` functions are asynchronous and you're using the existing value to update the next value. To do this pass in a callback function:

```
setViewAllLabel((prevProp) => prevProp === "View all" ? "View less" : "View all");
```

Do this for all the other states where you use its own value to update it.